src/jukebox/jukebox/plugs.py

```diff
@@ -0,0 +1,726 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+# Contributing author(s):
+# - Christian Banz
+
+"""
+A plugin package with some special functionality
+
+Plugins packages are python packages that are dynamically loaded. From these packages only a subset of objects is exposed
+through the plugs.call interface. The python packages can use decorators or dynamic function call to register (callable)
+objects.
+
+The python package name may be different from the name the package is registered under in plugs. This allows to load different
+python packages for a specific feature based on a configuration file. Note: Python package are still loaded as regular
+python packages and can be accessed by normal means
+
+If you want to provide additional functionality to the same feature (probably even for run-time switching)
+you can implement a Factory Pattern using this package. Take a look at volume.py as an example.
+
+Kick-start examples:
+    Example: Decorate a function for auto-registering under it's own name
+    >>> import jukebox.plugs as plugs
+    >>> @plugs.register
+    >>> def func1(param):
+    >>>     pass
+
+    Example: Decorate a function for auto-registering under a new name
+    >>> @plugs.register(name='better_name')
+    >>> def func2(param):
+    >>>     pass
+
+    Example: Register a function during run-time under it's own name
+    >>> def func3(param):
+    >>>     pass
+    >>> plugs.register(func3)
+
+    Example: Register a function during run-time under a new name
+    >>> def func4(param):
+    >>>     pass
+    >>> plugs.register(func4, name='other_name', package='other_package')
+
+    Example: Decorate a class for auto registering during initialization,
+    including all methods (see _register_class for more info)
+    >>> @plugs.register(auto_tag=True)
+    >>> class MyClass1:
+    >>>     pass
+
+    Example: Register a class instance, from which only report is a callable method through the plugs interface
+    >>> class MyClass2:
+    >>>     @plugs.tag
+    >>>     def report(self):
+    >>>         pass
+    >>> myinst2 = MyClass2()
+    >>> plugin.register(myinst2, name='myinst2')
+
+Naming convention:
+  package      : (a) Either a python package
+              : (b) or a plugin package (which si the python package but probably loaded under a different name inside plugs)
+  plugin       : (a) An object from the package that can be accessed through the plugs call function
+               :     (i.e. a function or a class instance)
+               : (b) the string name to above object
+  name         :     the string name of the plugin object for registration
```

Comment:
Here is a payload I created after this convention

```
{
  package: "player",
  plugin: "ctrl",
  name: "ctrl",
  method: "list_all_dirs",
  kwargs: {}
}
```

Will `plugin` and `name` always be the same? Do you have an example where those values are different

Reply:
A slight of misunderstanding here: Inside plugs.py, I need to differentiate between the python package name (plugin) and the alias name under which is well be accessible through plugs.call() / get() / exists() which may be the same or another (in this case "name"). From the outside there is only three elements: package, plugin, method. Attention: in this case the aliased name is used here. Take a look at plugs.call(). 
Also: When going through the RPC, the call signature is defined by the RPC. Which has the 3 elements package, plugin, method. (name is not used here - plugin is always the aliased name). There is bit of documentation in [rpc/server.py](https://github.com/ChisSoc/RPi-Jukebox-RFID/blob/0d57413fe55e908c824cd3baad890665700f3d58/src/jukebox/jukebox/rpc/server.py#L65):
```
            # {
            #   'package'  : str  # The plugin package loaded from python module
            #   'plugin'   : str  # The plugin object to be accessed from the package (i.e. function or class instance)
            #   'method'   : str  # (optional) The method of the class instance
            #   'args'     : [ ]  # (optional) Positional arguments as list
            #   'kwargs'   : { }  # (optional) Keyword arguments as dictionary
            #   'as_thread': bool # (optional) start call in separate thread
            #   'id'       : Any  # (optional) Round-trip id for response
            #   'tsp'      : Any  # (optional) measure and return total processing time for the call request
            # }

```
If you look at shared/settings/plugin_reference.txt it is
first column: package
second column: plugin
third column: method (if any / optional)